libs/my92xx/lightbulb_my92xx_driver.c

```diff
@@ -0,0 +1,232 @@
+/**
+ * my92xx LED Driver
+ * Based on the C++ driver by Xose Pérez
+ * Copyright (c) 2016 - 2026 MaiKe Labs
+ * Copyright (C) 2017 - 2018 Xose Pérez for the Arduino compatible library
+ *  
+ * This code is a C port of the my92xx LED Driver Library by Xose Perez
+ * the original C++ implementation is https://github.com/xoseperez/my92xx
+ * Some functions were adjusted but mainly it's a copy paste of the library mentioned above.
+ * Original comments are kept.
+ * 
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include "lightbulb_my92xx_driver.h"
+#include <FreeRTOS.h>
+#include <espressif/esp_common.h>
+#include <string.h>
+
+
+ #define MY92XX_DEFAULT_FREQUENCY  (2000)
+
+typedef struct _my92xx_channel {
+    uint16_t duty;
+    uint8_t channel;
+    struct _my92xx_channel* next;
+} my92xx_channel_t;
+
+typedef struct _my92xx_driver_info{
+    uint8_t di_pin;
+    uint8_t dcki_pin;
+    uint8_t numChips;
+    uint8_t numChannels;
+    my92xx_channel_t* channels;
+} my92xx_driver_info_t;
+
+void my92xx_send_pulses(uint8_t pin, unsigned int times);
+void my92xx_write(unsigned int data, unsigned char bit_length);
+void my92xx_send();
+
+static my92xx_cmd_t config = {
+    .bit_width = MY92XX_CMD_BIT_WIDTH_16,
+    .scatter = MY92XX_CMD_SCATTER_APDM,
+    .frequency = MY92XX_CMD_FREQUENCY_DIVIDE_1,
+    .reaction = MY92XX_CMD_REACTION_FAST,
+    .one_shot = MY92XX_CMD_ONE_SHOT_DISABLE,
+    .resv = 0
+};
```

Comment:
Not needed data struct. Waste of RAM.

Reply:
Do you suggest to allocate it dynamically or rather that one can avoid use of the struct and directly make byte array?